python/pyspark/sql/catalog.py

```diff
@@ -255,26 +255,67 @@ def registerFunction(self, name, f, returnType=StringType()):
         >>> _ = spark.udf.register("stringLengthInt", len, IntegerType())
         >>> spark.sql("SELECT stringLengthInt('test')").collect()
         [Row(stringLengthInt(test)=4)]
+        """
+
+        # This is to check whether the input function is a wrapped/native UserDefinedFunction
+        if hasattr(f, 'asNondeterministic'):
+            raise ValueError("Please use registerUDF for registering UDF. The expected function of "
+                             "registerFunction is a Python function (including lambda function)")
+        udf = UserDefinedFunction(f, returnType=returnType, name=name,
+                                  evalType=PythonEvalType.SQL_BATCHED_UDF)
+        self._jsparkSession.udf().registerPython(name, udf._judf)
+        return udf._wrapped()
+
+    @ignore_unicode_prefix
+    @since(2.3)
+    def registerUDF(self, name, f):
+        """Registers a :class:`UserDefinedFunction`. The registered UDF can be used in SQL
+        statement.
+
+        :param name: name of the UDF
```

Comment:
Maybe something like "name of the UDF in SQL statement" to be more specific.

Reply:
Actually, `name of the UDF in SQL statement` is wrong. The name is also used as the name of returned UDF.

Reply:
Hm .. actually I think we should not change the name of the returned UDF?

Reply:
Why?

Reply:
Because I thought it registers it to SQL statement with the given name and the defined UDF instance (`f`) shouldn't change its name itself at the register time.

Up to my knowledge, we happened to have `registerUDF` to avoid such problem in `returnType`.

Reply:
I also think changing the property (name) of the input udf object in `registerUDF` is unintuitive..

Reply:
After checking with our Scala API, I found the name is not changed. Let me update it.